### PR TITLE
Switch to Poetry as a dependency manager

### PR DESCRIPTION
**Pull request by lugi0**

### CODE REVIEW

**Comment by lugi0** [comment]
@jgarciao Responding to your comment in #688 

"
A few comments:

    1. It would be better to have ods-ci for the main folder (I think dash it's recommended over underscore for folder names)
    2. I believe merging this might break CI and possibly ods-ci docker image
    3. Can the docs folder be at the root of the repository?
    4. Can you add a new document docs/release.md with the instructions to create poetry.lock and the other steps to create an ods-ci release?
    5. Will the steps in Readme.md > Quick Start work or they need to be updated somehow?
"

  1. [Poetry requires](https://python-poetry.org/docs/basic-usage/#project-setup) the project to be structured in this way. We can change the name of the subfolder, but I would prefer not having ~/ods-ci/ods-ci/...
  2. Unsure about the docker image. It's possible CI needs to be updated to cd into ods_ci before running run_robot_tests. This is because the bash script runs some commands relative to the workind directory from where the file is run, so from the root trying to run `./ods_ci/run_robot_tests.sh` will fail.
  3. It can, although IMHO it shouldn't. It makes sense to me to keep the docs in the project folder.
  4. I sure can
  5. As long as Poetry is installed and available on the $PATH, run_robot_test will use it and so nothing else should change in the quick start

**Comment by lugi0** [comment]
@jgarciao The dockerfile was indeed broken, I've now updated it and confirmed that it works with Poetry. I've also updated its documentation to highlight the changes. 
I've added the `release.md` doc file to describe how to release ods-ci and how to update dependencies with Poetry

File: ods_ci/build/Dockerfile [hunk 1]
```diff
@@ -38,31 +38,34 @@ WORKDIR $HOME/ods-ci
...
-COPY utils/scripts/ocm/templates/create_ldap_idp.jinja utils/scripts/ocm/templates/create_ldap_idp.jinja
-COPY requirements.txt setup.py .
-RUN  chmod +x build/run.sh
-RUN  python3 --version
-RUN  python3 -m venv venv && source venv/bin/activate &&  pip3 install --upgrade pip && venv/bin/pip3 install -r requirements.txt
+COPY ods_ci/tests ods_ci/tests/
+COPY ods_ci/tasks ods_ci/tasks/
+COPY ods_ci/libs ods_ci/libs/
+COPY ods_ci/run_robot_test.sh ods_ci/run_robot_test.sh
+COPY ods_ci/build/run.sh  ods_ci/run.sh
```
Comment (bdattoma) [inline ods_ci/build/Dockerfile:45]: why removing the "build" folder?

File: ods_ci/build/Dockerfile [hunk 2]
```diff
@@ -38,31 +38,34 @@ WORKDIR $HOME/ods-ci
 RUN groupadd --gid 1001 ods-ci-users
 RUN useradd -r -u 1001 -g ods-ci-users ods-ci-runner
 
-COPY tests tests/
-COPY tasks tasks/
-COPY libs libs/
-COPY run_robot_test.sh  .
-COPY build/run.sh  build/run.sh
-COPY build/install_idp.sh  build/install_idp.sh
```
Comment (bdattoma) [inline ods_ci/build/Dockerfile:46]: I think this line got removed in this PR, but it should be present

File: ods_ci/build/Dockerfile [hunk 3]
```diff
@@ -38,31 +38,34 @@ WORKDIR $HOME/ods-ci
...
+COPY ods_ci/configs/resources/oauth_ldap_idp.json ods_ci/configs/resources/oauth_ldap_idp.json
+COPY ods_ci/configs/templates/ca-rolebinding.yaml ods_ci/configs/templates/ca-rolebinding.yaml
+COPY ods_ci/utils/scripts/ocm/templates/create_ldap_idp.jinja ods_ci/utils/scripts/ocm/templates/create_ldap_idp.jinja
+RUN  chmod +x ods_ci/run.sh
+COPY pyproject.toml .
+COPY poetry.lock .
+COPY README.md .
+RUN python3 --version
+RUN curl -sSL https://install.python-poetry.org | python3 -
+ENV PATH="${PATH}:$HOME/.local/bin"
```
[the remark this answers is not in the document]
Comment (lugi0) [reply ods_ci/build/Dockerfile:64]: Not really sure what you mean here, those lines use curl as well?

File: ods_ci/build/Dockerfile (hunk 2 above)
Comment (lugi0) [reply ods_ci/build/Dockerfile:46]: fixed

File: ods_ci/build/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/build/Dockerfile:65]: yes, but not overwriting the PATH variable

[the remark this answers is not in the document]
Comment (lugi0) [reply ods_ci/build/Dockerfile:65]: that's not overwriting the PATH, it's appending a new folder to it. That way we can use `poetry` as a command in the container

[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/build/Dockerfile:65]: yeah that's what I meant too xD you can move the poetry executable to usr/local/bin folder like done with oc and ocm

**Review by jgarciao — DISMISSED** [review]
lgtm

**Review by jgarciao — DISMISSED** [review]
lgtm

File: ods_ci/build/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (lugi0) [reply ods_ci/build/Dockerfile:65]: I'd rather not mess with the Poetry install as it has its own defaults for the virtual environments it creates and I'm not sure at this point what would happen when moving the exec file. 
I don't see the problem in adding a folder to the $PATH however, what was your concern here?

**Comment by aloganat** [comment]
lgtm

File: ods_ci/run_robot_test.sh [hunk 4]
```diff
@@ -287,17 +287,8 @@ if command -v yq &> /dev/null
...
-  python3 -m venv "${VENV_ROOT}"
-fi
-# shellcheck disable=SC1091,SC2086
-source ${VENV_ROOT}/bin/activate
-
-if [[ ${SKIP_PIP_INSTALL} -eq 0 ]]; then
-  ${VENV_ROOT}/bin/pip install --upgrade pip
-  ${VENV_ROOT}/bin/pip install -r requirements.txt
-fi
+poetry install
```
[the remark this answers is not in the document]
Comment (lugi0) [reply ods_ci/run_robot_test.sh:290]: Oh, that totally slipped through. I'll send a PR to re-enable the skip.

[the remark this answers is not in the document]
Comment (kpouget) [reply ods_ci/run_robot_test.sh:290]: thanks :)

[the remark this answers is not in the document]
Comment (lugi0) [reply ods_ci/run_robot_test.sh:290]: Here it is: https://github.com/red-hat-data-services/ods-ci/pull/726

**Pull request merged**